### PR TITLE
MULE-16574: Fix message propagation and variables scope in policies

### PR DESCRIPTION
**Pull request by balbifm**

### CODE REVIEW

File: core/src/main/java/org/mule/runtime/core/api/policy/PolicyChain.java [hunk 1]
```diff
@@ -35,7 +35,7 @@
 import org.mule.runtime.core.api.processor.strategy.ProcessingStrategy;
 import org.mule.runtime.core.internal.context.notification.DefaultFlowCallStack;
 import org.mule.runtime.core.internal.exception.MessagingException;
-import org.mule.runtime.core.internal.policy.PolicyNotificationHelper;
+import org.mule.runtime.core.api.policy.PolicyNotificationHelper;
```
Comment (sonar-mulesoft) [inline core/src/main/java/org/mule/runtime/core/api/policy/PolicyChain.java]: Code Smell Code Smell: Remove this unnecessary import: same package classes are always implicitly imported. ([squid:UselessImportCheck](https://sonarqube.kbuild.msap.io/coding_rules#rule_key=squid%3AUselessImportCheck))

[See it in SonarQube](https://sonarqube.kbuild.msap.io/project/issues?id=mule-runtime-mule&pullRequest=7760&issues=AWnGD4_aotIXnmsUaS7h&open=AWnGD4_aotIXnmsUaS7h)

[ISSUE_KEY]: AWnGD4_aotIXnmsUaS7h

File: core/src/main/java/org/mule/runtime/core/internal/policy/AbstractCompositePolicy.java [hunk 2]
```diff
@@ -13,6 +13,7 @@
 import org.mule.runtime.api.exception.MuleException;
 import org.mule.runtime.core.api.event.CoreEvent;
 import org.mule.runtime.core.api.policy.Policy;
+import org.mule.runtime.core.api.policy.PolicyNotificationHelper;
```
Comment (sonar-mulesoft) [inline core/src/main/java/org/mule/runtime/core/internal/policy/AbstractCompositePolicy.java]: Code Smell Code Smell: Remove this unused import 'org.mule.runtime.core.api.policy.PolicyNotificationHelper'. ([squid:UselessImportCheck](https://sonarqube.kbuild.msap.io/coding_rules#rule_key=squid%3AUselessImportCheck))

[See it in SonarQube](https://sonarqube.kbuild.msap.io/project/issues?id=mule-runtime-mule&pullRequest=7760&issues=AWnGD5CGotIXnmsUaS7j&open=AWnGD5CGotIXnmsUaS7j)

[ISSUE_KEY]: AWnGD5CGotIXnmsUaS7j

File: core/src/main/java/org/mule/runtime/core/internal/policy/SourcePolicyProcessor.java [hunk 3]
```diff
@@ -8,7 +8,6 @@
 
 import static org.mule.runtime.core.internal.policy.PolicyNextActionMessageProcessor.POLICY_IS_PROPAGATE_MESSAGE_TRANSFORMATIONS;
 import static org.mule.runtime.core.internal.policy.PolicyNextActionMessageProcessor.POLICY_NEXT_OPERATION;
-import static org.mule.runtime.core.internal.policy.PolicyNextActionMessageProcessor.POLICY_STATE_EVENT;
 import static reactor.core.publisher.Flux.from;
 
 import org.mule.runtime.api.exception.MuleException;
```
Comment (sonar-mulesoft) [inline core/src/main/java/org/mule/runtime/core/internal/policy/SourcePolicyProcessor.java]: Code Smell Code Smell: Remove this unused import 'org.mule.runtime.api.exception.MuleException'. ([squid:UselessImportCheck](https://sonarqube.kbuild.msap.io/coding_rules#rule_key=squid%3AUselessImportCheck))

[See it in SonarQube](https://sonarqube.kbuild.msap.io/project/issues?id=mule-runtime-mule&pullRequest=7760&issues=AWnGD5BWotIXnmsUaS7i&open=AWnGD5BWotIXnmsUaS7i)

[ISSUE_KEY]: AWnGD5BWotIXnmsUaS7i

File: core/src/main/java/org/mule/runtime/core/internal/policy/CompositeSourcePolicy.java [hunk 4]
```diff
@@ -165,30 +149,43 @@ public CompositeSourcePolicy(List<Policy> parameterizedPolicies,
...
+    return componentLocation.getParts().get(0).getPartPath() + "/" + componentLocation.getParts().get(1).getPartPath()
+        + "[after next]";
+  }
+
+  private boolean isEventContextHandledByThisNext(CoreEvent event) {
+    final Set<String> eventCtxIds = ((InternalEvent) event).getInternalParameter(POLICY_NEXT_EVENT_CTX_IDS);
+    return eventCtxIds != null && eventCtxIds.contains(event.getContext().getId());
+  }
+
+  private InternalEvent mapErrorInInnerExecution(MessagingException messagingException) {
```
Comment (sonar-mulesoft) [inline core/src/main/java/org/mule/runtime/core/internal/policy/CompositeSourcePolicy.java]: Code Smell Code Smell: Remove this unused private "mapErrorInInnerExecution" method. ([squid:UnusedPrivateMethod](https://sonarqube.kbuild.msap.io/coding_rules#rule_key=squid%3AUnusedPrivateMethod))

[See it in SonarQube](https://sonarqube.kbuild.msap.io/project/issues?id=mule-runtime-mule&pullRequest=7760&issues=AWnGD5CXotIXnmsUaS7k&open=AWnGD5CXotIXnmsUaS7k)

[ISSUE_KEY]: AWnGD5CXotIXnmsUaS7k

File: core/src/main/java/org/mule/runtime/core/internal/policy/CompositeSourcePolicy.java [hunk 5]
```diff
@@ -239,15 +228,14 @@ private MessagingException mapErrorInInnerExecution(MessagingException messaging
...
-   * @throws Exception if there was an unexpected failure thrown by executing the chain.
    */
   @Override
   public Publisher<Either<SourcePolicyFailureResult, SourcePolicySuccessResult>> process(CoreEvent sourceEvent,
                                                                                          MessageSourceResponseParametersProcessor respParamProcessor) {
     return commonPolicy.process(sourceEvent, respParamProcessor);
   }
 
-  protected Supplier<Map<String, Object>> resolveErrorResponseParameters(MessagingException e) {
+  private Supplier<Map<String, Object>> resolveErrorResponseParameters(MessagingException e) {
```
Comment (sonar-mulesoft) [inline core/src/main/java/org/mule/runtime/core/internal/policy/CompositeSourcePolicy.java]: Code Smell Code Smell: Move this method into "SourceWithPoliciesFluxObjectFactory". ([squid:S3398](https://sonarqube.kbuild.msap.io/coding_rules#rule_key=squid%3AS3398))

[See it in SonarQube](https://sonarqube.kbuild.msap.io/project/issues?id=mule-runtime-mule&pullRequest=7760&issues=AWnGD5CXotIXnmsUaS7l&open=AWnGD5CXotIXnmsUaS7l)

[ISSUE_KEY]: AWnGD5CXotIXnmsUaS7l

File: core/src/main/java/org/mule/runtime/core/internal/policy/AbstractCompositePolicy.java (hunk 2 above)
Comment (sonar-mulesoft) [inline core/src/main/java/org/mule/runtime/core/internal/policy/AbstractCompositePolicy.java]: Code Smell Code Smell: Remove this unused import 'org.mule.runtime.core.api.policy.PolicyNotificationHelper'. ([squid:UselessImportCheck](https://sonarqube.kbuild.msap.io/coding_rules#rule_key=squid%3AUselessImportCheck))

[See it in SonarQube](https://sonarqube.kbuild.msap.io/project/issues?id=mule-runtime-mule&pullRequest=7760&issues=AWnGD5CGotIXnmsUaS7j&open=AWnGD5CGotIXnmsUaS7j)

[ISSUE_KEY]: AWnGD5CGotIXnmsUaS7j

File: core/src/main/java/org/mule/runtime/core/api/policy/PolicyChain.java [hunk 6]
```diff
@@ -129,6 +129,10 @@ public CoreEvent process(CoreEvent event) throws MuleException {
         .doOnNext(e -> notificationHelper.fireNotification(e, null, PROCESS_END));
   }
 
+  public MuleContext getMuleContext() {
```
Comment (sonar-mulesoft) [inline core/src/main/java/org/mule/runtime/core/api/policy/PolicyChain.java]: Code Smell Code Smell: Remove this use of "MuleContext"; it is deprecated. ([squid:CallToDeprecatedMethod](https://sonarqube.kbuild.msap.io/coding_rules#rule_key=squid%3ACallToDeprecatedMethod))

[See it in SonarQube](https://sonarqube.kbuild.msap.io/project/issues?id=mule-runtime-mule&pullRequest=7760&issues=AWnHtdUOotIXnmsUhQGB&open=AWnHtdUOotIXnmsUhQGB)

[ISSUE_KEY]: AWnHtdUOotIXnmsUhQGB

File: core/src/main/java/org/mule/runtime/core/internal/policy/PolicyNextActionMessageProcessor.java [hunk 7]
```diff
@@ -58,113 +51,73 @@
...
-
-          if (isEventContextHandledByThisNext(event)) {
-            MessagingException me = (MessagingException) error;
-            notificationHelper.fireNotification(me.getEvent(), me, AFTER_NEXT);
-            pushAfterNextFlowStackElement().accept(me.getEvent());
-
-            me.setProcessedEvent(policyEventConverter.createEvent((PrivilegedEvent) me.getEvent(),
-                                                                  loadState((PrivilegedEvent) me.getEvent())));
+        .map(result -> policyEventMapper.fromPolicyNext(result))
+        .onErrorContinue(MessagingException.class, (error, v) -> {
```
Comment (sonar-mulesoft) [inline core/src/main/java/org/mule/runtime/core/internal/policy/PolicyNextActionMessageProcessor.java]: Code Smell Code Smell: Remove useless curly braces around statement ([squid:S1602](https://sonarqube.kbuild.msap.io/coding_rules#rule_key=squid%3AS1602))

[See it in SonarQube](https://sonarqube.kbuild.msap.io/project/issues?id=mule-runtime-mule&pullRequest=7760&issues=AWnHtdTcotIXnmsUhQF-&open=AWnHtdTcotIXnmsUhQF-)

[ISSUE_KEY]: AWnHtdTcotIXnmsUhQF-

File: core/src/main/java/org/mule/runtime/core/internal/policy/PolicyNextActionMessageProcessor.java [hunk 8]
```diff
@@ -27,24 +24,20 @@
...
-import org.mule.runtime.core.internal.message.InternalEvent.Builder;
-import org.mule.runtime.core.privileged.event.BaseEventContext;
-import org.mule.runtime.core.privileged.event.PrivilegedEvent;
-
-import org.reactivestreams.Publisher;
-import org.slf4j.Logger;
-
-import com.google.common.collect.ImmutableSet;
 
 import java.util.Set;
```
Comment (sonar-mulesoft) [inline core/src/main/java/org/mule/runtime/core/internal/policy/PolicyNextActionMessageProcessor.java]: Code Smell Code Smell: Remove this unused import 'java.util.Set'. ([squid:UselessImportCheck](https://sonarqube.kbuild.msap.io/coding_rules#rule_key=squid%3AUselessImportCheck))

[See it in SonarQube](https://sonarqube.kbuild.msap.io/project/issues?id=mule-runtime-mule&pullRequest=7760&issues=AWnHtdTcotIXnmsUhQF_&open=AWnHtdTcotIXnmsUhQF_)

[ISSUE_KEY]: AWnHtdTcotIXnmsUhQF_

File: core/src/main/java/org/mule/runtime/core/internal/policy/CompositeSourcePolicy.java [hunk 9]
```diff
@@ -239,15 +184,14 @@ private MessagingException mapErrorInInnerExecution(MessagingException messaging
...
```
Comment (sonar-mulesoft) [inline core/src/main/java/org/mule/runtime/core/internal/policy/CompositeSourcePolicy.java]: Code Smell Code Smell: Move this method into "SourceWithPoliciesFluxObjectFactory". ([squid:S3398](https://sonarqube.kbuild.msap.io/coding_rules#rule_key=squid%3AS3398))

[See it in SonarQube](https://sonarqube.kbuild.msap.io/project/issues?id=mule-runtime-mule&pullRequest=7760&issues=AWnGD5CXotIXnmsUaS7l&open=AWnGD5CXotIXnmsUaS7l)

[ISSUE_KEY]: AWnGD5CXotIXnmsUaS7l

File: core/src/main/java/org/mule/runtime/core/api/policy/PolicyChain.java (hunk 1 above)
Comment (sonar-mulesoft) [inline core/src/main/java/org/mule/runtime/core/api/policy/PolicyChain.java]: Code Smell Code Smell: Remove this unnecessary import: same package classes are always implicitly imported. ([squid:UselessImportCheck](https://sonarqube.kbuild.msap.io/coding_rules#rule_key=squid%3AUselessImportCheck))

[See it in SonarQube](https://sonarqube.kbuild.msap.io/project/issues?id=mule-runtime-mule&pullRequest=7760&issues=AWnGD4_aotIXnmsUaS7h&open=AWnGD4_aotIXnmsUaS7h)

[ISSUE_KEY]: AWnGD4_aotIXnmsUaS7h

File: core/src/main/java/org/mule/runtime/core/internal/policy/PolicyNextActionMessageProcessor.java [hunk 10]
```diff
@@ -27,24 +24,20 @@
 import org.mule.runtime.core.api.MuleContext;
 import org.mule.runtime.core.api.context.notification.FlowStackElement;
 import org.mule.runtime.core.api.event.CoreEvent;
+import org.mule.runtime.core.api.policy.PolicyNotificationHelper;
 import org.mule.runtime.core.api.processor.Processor;
 import org.mule.runtime.core.internal.context.notification.DefaultFlowCallStack;
 import org.mule.runtime.core.internal.exception.MessagingException;
 import org.mule.runtime.core.internal.message.InternalEvent;
```
Comment (sonar-mulesoft) [inline core/src/main/java/org/mule/runtime/core/internal/policy/PolicyNextActionMessageProcessor.java]: Code Smell Code Smell: Remove this unused import 'org.mule.runtime.core.internal.message.InternalEvent'. ([squid:UselessImportCheck](https://sonarqube.kbuild.msap.io/coding_rules#rule_key=squid%3AUselessImportCheck))

[See it in SonarQube](https://sonarqube.kbuild.msap.io/project/issues?id=mule-runtime-mule&pullRequest=7760&issues=AWnHtdTcotIXnmsUhQGA&open=AWnHtdTcotIXnmsUhQGA)

[ISSUE_KEY]: AWnHtdTcotIXnmsUhQGA

File: core/src/main/java/org/mule/runtime/core/internal/policy/SourcePolicyProcessor.java (hunk 3 above)
Comment (sonar-mulesoft) [inline core/src/main/java/org/mule/runtime/core/internal/policy/SourcePolicyProcessor.java]: Code Smell Code Smell: Remove this unused import 'org.mule.runtime.api.exception.MuleException'. ([squid:UselessImportCheck](https://sonarqube.kbuild.msap.io/coding_rules#rule_key=squid%3AUselessImportCheck))

[See it in SonarQube](https://sonarqube.kbuild.msap.io/project/issues?id=mule-runtime-mule&pullRequest=7760&issues=AWnGD5BWotIXnmsUaS7i&open=AWnGD5BWotIXnmsUaS7i)

[ISSUE_KEY]: AWnGD5BWotIXnmsUaS7i

Comment (sonar-mulesoft) [inline core/src/main/java/org/mule/runtime/core/internal/policy/SourcePolicyProcessor.java]: Code Smell Code Smell: Remove this unused import 'org.mule.runtime.api.exception.MuleException'. ([squid:UselessImportCheck](https://sonarqube.kbuild.msap.io/coding_rules#rule_key=squid%3AUselessImportCheck))

[See it in SonarQube](https://sonarqube.kbuild.msap.io/project/issues?id=mule-runtime-mule&pullRequest=7760&issues=AWnGD5BWotIXnmsUaS7i&open=AWnGD5BWotIXnmsUaS7i)

[ISSUE_KEY]: AWnGD5BWotIXnmsUaS7i

File: core/src/main/java/org/mule/runtime/core/api/policy/PolicyChain.java (hunk 6 above)
Comment (sonar-mulesoft) [inline core/src/main/java/org/mule/runtime/core/api/policy/PolicyChain.java]: Code Smell Code Smell: Remove this use of "MuleContext"; it is deprecated. ([squid:CallToDeprecatedMethod](https://sonarqube.kbuild.msap.io/coding_rules#rule_key=squid%3ACallToDeprecatedMethod))

[See it in SonarQube](https://sonarqube.kbuild.msap.io/project/issues?id=mule-runtime-mule&pullRequest=7760&issues=AWnHtdUOotIXnmsUhQGB&open=AWnHtdUOotIXnmsUhQGB)

[ISSUE_KEY]: AWnHtdUOotIXnmsUhQGB

File: core/src/main/java/org/mule/runtime/core/internal/policy/CompositeSourcePolicy.java [hunk 11]
```diff
@@ -239,25 +195,13 @@ private MessagingException mapErrorInInnerExecution(MessagingException messaging
...
-        ((InternalEvent) e.getEvent()).getInternalParameter(POLICY_SOURCE_ORIGINAL_FAILURE_RESPONSE_PARAMETERS);
-
-    return () -> getParametersTransformer()
-        .map(parametersTransformer -> concatMaps(originalFailureResponseParameters,
-                                                 parametersTransformer
-                                                     .fromMessageToErrorResponseParameters(e.getEvent().getMessage())))
-        .orElse(originalFailureResponseParameters);
-  }
-
   private Map<String, Object> concatMaps(Map<String, Object> originalResponseParameters,
```
Comment (sonar-mulesoft) [inline core/src/main/java/org/mule/runtime/core/internal/policy/CompositeSourcePolicy.java]: Code Smell Code Smell: Move this method into "SourceWithPoliciesFluxObjectFactory". ([squid:S3398](https://sonarqube.kbuild.msap.io/coding_rules#rule_key=squid%3AS3398))

[See it in SonarQube](https://sonarqube.kbuild.msap.io/project/issues?id=mule-runtime-mule&pullRequest=7760&issues=AWnJ1eXkotIXnmsUjAK5&open=AWnJ1eXkotIXnmsUjAK5)

[ISSUE_KEY]: AWnJ1eXkotIXnmsUjAK5

File: core/src/main/java/org/mule/runtime/core/internal/policy/PolicyNextActionMessageProcessor.java (hunk 7 above)
Comment (sonar-mulesoft) [inline core/src/main/java/org/mule/runtime/core/internal/policy/PolicyNextActionMessageProcessor.java]: Code Smell Code Smell: Remove useless curly braces around statement ([squid:S1602](https://sonarqube.kbuild.msap.io/coding_rules#rule_key=squid%3AS1602))

[See it in SonarQube](https://sonarqube.kbuild.msap.io/project/issues?id=mule-runtime-mule&pullRequest=7760&issues=AWnHtdTcotIXnmsUhQF-&open=AWnHtdTcotIXnmsUhQF-)

[ISSUE_KEY]: AWnHtdTcotIXnmsUhQF-

File: core/src/main/java/org/mule/runtime/core/api/policy/PolicyChain.java (hunk 1 above)
Comment (sonar-mulesoft) [inline core/src/main/java/org/mule/runtime/core/api/policy/PolicyChain.java]: Code Smell Code Smell: Remove this unnecessary import: same package classes are always implicitly imported. ([squid:UselessImportCheck](https://sonarqube.kbuild.msap.io/coding_rules#rule_key=squid%3AUselessImportCheck))

[See it in SonarQube](https://sonarqube.kbuild.msap.io/project/issues?id=mule-runtime-mule&pullRequest=7760&issues=AWnGD4_aotIXnmsUaS7h&open=AWnGD4_aotIXnmsUaS7h)

[ISSUE_KEY]: AWnGD4_aotIXnmsUaS7h

File: core/src/main/java/org/mule/runtime/core/internal/policy/PolicyNextActionMessageProcessor.java (hunk 8 above)
Comment (sonar-mulesoft) [inline core/src/main/java/org/mule/runtime/core/internal/policy/PolicyNextActionMessageProcessor.java]: Code Smell Code Smell: Remove this unused import 'java.util.Set'. ([squid:UselessImportCheck](https://sonarqube.kbuild.msap.io/coding_rules#rule_key=squid%3AUselessImportCheck))

[See it in SonarQube](https://sonarqube.kbuild.msap.io/project/issues?id=mule-runtime-mule&pullRequest=7760&issues=AWnHtdTcotIXnmsUhQF_&open=AWnHtdTcotIXnmsUhQF_)

[ISSUE_KEY]: AWnHtdTcotIXnmsUhQF_

File: core/src/main/java/org/mule/runtime/core/internal/policy/AbstractCompositePolicy.java (hunk 2 above)
Comment (sonar-mulesoft) [inline core/src/main/java/org/mule/runtime/core/internal/policy/AbstractCompositePolicy.java]: Code Smell Code Smell: Remove this unused import 'org.mule.runtime.core.api.policy.PolicyNotificationHelper'. ([squid:UselessImportCheck](https://sonarqube.kbuild.msap.io/coding_rules#rule_key=squid%3AUselessImportCheck))

[See it in SonarQube](https://sonarqube.kbuild.msap.io/project/issues?id=mule-runtime-mule&pullRequest=7760&issues=AWnGD5CGotIXnmsUaS7j&open=AWnGD5CGotIXnmsUaS7j)

[ISSUE_KEY]: AWnGD5CGotIXnmsUaS7j

File: core/src/main/java/org/mule/runtime/core/internal/policy/PolicyNextActionMessageProcessor.java (hunk 10 above)
Comment (sonar-mulesoft) [inline core/src/main/java/org/mule/runtime/core/internal/policy/PolicyNextActionMessageProcessor.java]: Code Smell Code Smell: Remove this unused import 'org.mule.runtime.core.internal.message.InternalEvent'. ([squid:UselessImportCheck](https://sonarqube.kbuild.msap.io/coding_rules#rule_key=squid%3AUselessImportCheck))

[See it in SonarQube](https://sonarqube.kbuild.msap.io/project/issues?id=mule-runtime-mule&pullRequest=7760&issues=AWnHtdTcotIXnmsUhQGA&open=AWnHtdTcotIXnmsUhQGA)

[ISSUE_KEY]: AWnHtdTcotIXnmsUhQGA

File: core/src/main/java/org/mule/runtime/core/internal/policy/AbstractCompositePolicy.java [hunk 12]
```diff
@@ -51,22 +51,15 @@ public AbstractCompositePolicy(List<Policy> policies,
...
-  /**
-   * When this policy is processed, it will use the {@link CompositeOperationPolicy.NextOperationCall} which will keep track of
-   * the different policies to be applied and the current index of the policy under execution.
-   * <p>
-   * The first time, the first policy in the {@code parameterizedPolicies} will be executed, it will receive as it next operation
-   * the same instance of {@link CompositeOperationPolicy.NextOperationCall}, and since
-   * {@link CompositeOperationPolicy.NextOperationCall} keeps track of the policy executed, it will execute the following policy
-   * in the chain until the finally policy it's executed in which case then next operation of it, it will be the operation
-   * execution.
-   */
```
Comment (mbuchwald) [inline core/src/main/java/org/mule/runtime/core/internal/policy/AbstractCompositePolicy.java]: why did you remove the javadoc?

File: core-tests/src/test/java/org/mule/runtime/core/internal/policy/AbstractPolicyProcessorTestCase.java [hunk 13]
```diff
@@ -81,32 +82,7 @@ public void before() {
...
-  public void variablesAddedBeforeNextProcessorNotPropagatedToIt() throws MuleException {
-    CoreEvent initialEventWithVars = CoreEvent.builder(initialEvent).addVariable(INIT_VAR_NAME, INIT_VAR_VALUE).build();
-    when(policy.getPolicyChain().apply(any())).thenAnswer(invocation -> subscriberContext()
-        .flatMap(ctx -> Mono.<CoreEvent>from(invocation.getArgument(0)).transform(ctx.get(POLICY_NEXT_OPERATION))));
-
-    just(initialEventWithVars).transform(policyProcessor).block();
-
-    verify(flowProcessor).apply(eventCaptor.capture());
-    assertEquals(((CoreEvent) from(eventCaptor.getValue()).block()).getVariables().keySet(),
-                 initialEventWithVars.getVariables().keySet());
```
Comment (mbuchwald) [inline core-tests/src/test/java/org/mule/runtime/core/internal/policy/AbstractPolicyProcessorTestCase.java]: why are this tests being removed?

File: core-tests/src/test/java/org/mule/runtime/core/internal/policy/CompositeSourcePolicyTestCase.java [hunk 14]
```diff
@@ -227,33 +225,6 @@ public void policyExecutionFailureAfterNextPropagates() throws Exception {
...
-    compositeSourcePolicy =
-        new CompositeSourcePolicy(asList(firstPolicy, secondPolicy), flowExecutionProcessor,
-                                  of(sourcePolicyParametersTransformer), sourcePolicyProcessorFactory);
-
-    Either<SourcePolicyFailureResult, SourcePolicySuccessResult> sourcePolicyResult =
-        from(compositeSourcePolicy.process(initialEvent, sourceParametersProcessor)).block();
-    assertThat(sourcePolicyResult.isLeft(), is(true));
-    assertThat(sourcePolicyResult.getLeft().getMessagingException().getCause(), is(policyException));
-    verify(sourcePolicyParametersTransformer).fromFailureResponseParametersToMessage(any());
-    verify(sourcePolicyParametersTransformer, never()).fromSuccessResponseParametersToMessage(any());
```
Comment (mbuchwald) [inline core-tests/src/test/java/org/mule/runtime/core/internal/policy/CompositeSourcePolicyTestCase.java]: same here

File: core/src/main/java/org/mule/runtime/core/internal/policy/PolicyNextActionMessageProcessor.java [hunk 15]
```diff
@@ -58,113 +48,73 @@
...
-
-          if (isEventContextHandledByThisNext(event)) {
-            MessagingException me = (MessagingException) error;
-            notificationHelper.fireNotification(me.getEvent(), me, AFTER_NEXT);
-            pushAfterNextFlowStackElement().accept(me.getEvent());
-
-            me.setProcessedEvent(policyEventConverter.createEvent((PrivilegedEvent) me.getEvent(),
-                                                                  loadState((PrivilegedEvent) me.getEvent())));
+        .map(result -> policyEventMapper.fromPolicyNext(result))
+        .onErrorContinue(MessagingException.class, (error, v) -> {
```
Comment (sonar-mulesoft) [inline core/src/main/java/org/mule/runtime/core/internal/policy/PolicyNextActionMessageProcessor.java]: Code Smell Code Smell: Remove useless curly braces around statement ([squid:S1602](https://sonarqube.kbuild.msap.io/coding_rules#rule_key=squid%3AS1602))

[See it in SonarQube](https://sonarqube.kbuild.msap.io/project/issues?id=mule-runtime-mule&pullRequest=7760&issues=AWnHtdTcotIXnmsUhQF-&open=AWnHtdTcotIXnmsUhQF-)

[ISSUE_KEY]: AWnHtdTcotIXnmsUhQF-

File: core/src/main/java/org/mule/runtime/core/api/policy/PolicyChain.java (hunk 6 above)
Comment (sonar-mulesoft) [inline core/src/main/java/org/mule/runtime/core/api/policy/PolicyChain.java]: Code Smell Code Smell: Remove this use of "MuleContext"; it is deprecated. ([squid:CallToDeprecatedMethod](https://sonarqube.kbuild.msap.io/coding_rules#rule_key=squid%3ACallToDeprecatedMethod))

[See it in SonarQube](https://sonarqube.kbuild.msap.io/project/issues?id=mule-runtime-mule&pullRequest=7760&issues=AWnHtdUOotIXnmsUhQGB&open=AWnHtdUOotIXnmsUhQGB)

[ISSUE_KEY]: AWnHtdUOotIXnmsUhQGB

File: core/src/main/java/org/mule/runtime/core/internal/policy/AbstractCompositePolicy.java [hunk 16]
```diff
@@ -27,11 +27,10 @@
  * Abstract implementation that performs the chaining of a set of policies and the {@link Processor} being intercepted.
  *
  * @param <ParametersTransformer> the type of the function parameters transformer.
- * @param <ParametersProcessor> the type of the parameters processor that provides access to the initial value of the parameters.
  *
  * @since 4.0
  */
-public abstract class AbstractCompositePolicy<ParametersTransformer, Subject> {
+public abstract class AbstractCompositePolicy<ParametersTransformer> {
```
Comment (sonar-mulesoft) [inline core/src/main/java/org/mule/runtime/core/internal/policy/AbstractCompositePolicy.java]: Code Smell Code Smell: Rename this generic name to match the regular expression '^[A-Z][0-9]?$'. ([squid:S00119](https://sonarqube.kbuild.msap.io/coding_rules#rule_key=squid%3AS00119))

[See it in SonarQube](https://sonarqube.kbuild.msap.io/project/issues?id=mule-runtime-mule&pullRequest=7760&issues=AWnLDJg_otIXnmsUm4Yh&open=AWnLDJg_otIXnmsUm4Yh)

[ISSUE_KEY]: AWnLDJg_otIXnmsUm4Yh

File: core/src/main/java/org/mule/runtime/core/internal/policy/SourcePolicyProcessor.java (hunk 3 above)
Comment (sonar-mulesoft) [inline core/src/main/java/org/mule/runtime/core/internal/policy/SourcePolicyProcessor.java]: Code Smell Code Smell: Remove this unused import 'org.mule.runtime.api.exception.MuleException'. ([squid:UselessImportCheck](https://sonarqube.kbuild.msap.io/coding_rules#rule_key=squid%3AUselessImportCheck))

[See it in SonarQube](https://sonarqube.kbuild.msap.io/project/issues?id=mule-runtime-mule&pullRequest=7760&issues=AWnGD5BWotIXnmsUaS7i&open=AWnGD5BWotIXnmsUaS7i)

[ISSUE_KEY]: AWnGD5BWotIXnmsUaS7i

File: core/src/main/java/org/mule/runtime/core/internal/policy/CompositeSourcePolicy.java [hunk 17]
```diff
@@ -239,25 +194,13 @@ private MessagingException mapErrorInInnerExecution(MessagingException messaging
...
```
Comment (sonar-mulesoft) [inline core/src/main/java/org/mule/runtime/core/internal/policy/CompositeSourcePolicy.java]: Code Smell Code Smell: Move this method into "SourceWithPoliciesFluxObjectFactory". ([squid:S3398](https://sonarqube.kbuild.msap.io/coding_rules#rule_key=squid%3AS3398))

[See it in SonarQube](https://sonarqube.kbuild.msap.io/project/issues?id=mule-runtime-mule&pullRequest=7760&issues=AWnJ1eXkotIXnmsUjAK5&open=AWnJ1eXkotIXnmsUjAK5)

[ISSUE_KEY]: AWnJ1eXkotIXnmsUjAK5
